akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```diff
@@ -0,0 +1,538 @@
+     * Without this flattening a final pointless `map` step would be added
+     * for each iteration, eventually leading to an OutOfMemoryError.
+     */
+    def map[T, Mapped](f: Out ⇒ T)(implicit ev: MapAdapter[S, T, Mapped]): Operation[S, Mapped] = flatMap(ev.lift(f))
+
+    /**
+     * Unfortunately for-comprehensions desugar to `.foreach` if the `yield`
+     * keyword is absent, but they do so throughout the whole expression.
+     * Coding a normal signature for `foreach` would break everything after
+     * the first step as it would not be executed—the `Unit` result is there
+     * already.
+     */
+    def foreach[T, Mapped](f: Out ⇒ T)(implicit ev: MapAdapter[S, T, Mapped]): Operation[S, Mapped] = flatMap(ev.lift(f))
+
+    /**
+     * Only continue this process if the given predicate is fulfilled, terminate
+     * it otherwise.
+     */
+    def filter(p: Out ⇒ Boolean): Operation[S, Out] = flatMap(o ⇒ if (p(o)) Return(o) else ShortCircuit)
+
+    /**
+     * Only continue this process if the given predicate is fulfilled, terminate
+     * it otherwise.
+     */
+    def withFilter(p: Out ⇒ Boolean): Operation[S, Out] = flatMap(o ⇒ if (p(o)) Return(o) else ShortCircuit)
+
+    /**
+     * Wrap as a [[Process]] with infinite timeout and a mailbox capacity of 1.
+     * Small processes that are called or chained often interact in a fully
+     * sequential fashion, where these defaults make sense.
+     */
+    def named(name: String): Process[S, Out] = Process(name, Duration.Inf, 1, this)
+
+    /**
+     * Wrap as a [[Process]] with the given mailbox capacity and infinite timeout.
+     */
+    def withMailboxCapacity(mailboxCapacity: Int): Process[S, Out] = named("").withMailboxCapacity(mailboxCapacity)
+
+    /**
+     * Wrap as a [[Process]] with the given timeout and a mailbox capacity of 1.
+     */
+    def withTimeout(timeout: Duration): Process[S, Out] = named("").withTimeout(timeout)
+
+    /**
+     * Wrap as a [[Process]] but without a name and convert to a [[Behavior]].
+     */
+    def toBehavior: Behavior[ActorCmd[S]] = named("").toBehavior
+
+  }
+
+  /*
+   * These are the private values that make up the core algebra.
+   */
+
+  private[typed] case class FlatMap[S, Out1, Out2](first: Operation[S, Out1], then: Out1 ⇒ Operation[S, Out2]) extends Operation[S, Out2] {
+    override def toString: String = s"FlatMap($first)"
+  }
+  private[typed] case object ShortCircuit extends Operation[Nothing, Nothing] {
+    override def flatMap[T](f: Nothing ⇒ Operation[Nothing, T]): Operation[Nothing, T] = this
+  }
+
+  private[typed] case object System extends Operation[Nothing, ActorSystem[Nothing]]
+  private[typed] case object Read extends Operation[Nothing, Nothing]
+  private[typed] case object ProcessSelf extends Operation[Nothing, ActorRef[Any]]
+  private[typed] case object ActorSelf extends Operation[Nothing, ActorRef[ActorCmd[Nothing]]]
+  private[typed] case class Return[T](value: T) extends Operation[Nothing, T]
```

Comment:
the pattern matcher takes advantage of it (not sure exactly what the difference is)